internal/provider/host_resource.go

```diff
@@ -0,0 +1,390 @@
+package provider
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource              = &HostResource{}
+	_ resource.ResourceWithConfigure = &HostResource{}
+)
+
+// NewHostResource is a helper function to simplify the provider implementation
+func NewHostResource() resource.Resource {
+	return &HostResource{}
+}
+
+type HostResourceModelInterface interface {
+	ParseHttpResponse(body []byte) error
+	CreateRequestBody() ([]byte, diag.Diagnostics)
+	GetURL() string
+}
+
+// HostResource is the resource implementation.
+type HostResource struct {
+	client ProviderHTTPClient
+}
+
+// Metadata returns the resource type name.
+func (r *HostResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_host"
+}
+
+// Schema defines the schema for the host resource
+func (r *HostResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+
+			"inventory_id": schema.Int64Attribute{
+				Required: true,
+			},
+			"instance_id": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+			},
+			"name": schema.StringAttribute{
+				Required: true,
+			},
+			"description": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+			},
+			"host_url": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"variables": schema.StringAttribute{
+				Optional: true,
+			},
+			"enabled": schema.BoolAttribute{
+				Optional:    true,
+				Computed:    true,
+				Description: "Defaults true.",
+			},
+			"group_id": schema.Int64Attribute{
+				Optional:    true,
+				Description: "Set this option to associate an existing group with a host.",
+			},
+			"disassociate_group": schema.BoolAttribute{
```

Comment:
A group is not removed from a host unless disassociate parameter is passed. See https://docs.ansible.com/automation-controller/latest/html/controllerapi/api_ref.html#/Hosts/Hosts_hosts_groups_create Also, id must be passed as an integer and a list is not accepted, this is my understanding.

I can implement something like:

``` 
 groups = [
    {
      "id" : 2
    },
    {
      "id": 3
      "disassociate": true
    }
  ]
}``` 

It will mainly make an group association for every id.

Reply:
We need to do the work of figuring out which groups need to be added/removed on our side. We'll need to list the groups a host is a member of, compare that to what's in the tf config, and then add and remove as necessary. The terraform config should look something like:

```
resource "aap_host" "foo_host" {
  ...
  groups = [group_a.id, group_b.id]
}

resource "aap_group" "group_a" {}
resource "aap_group" "group_b" {}
```